### PR TITLE
feat: reproducible choice interactive

### PR DESCRIPTION
**Pull request by dj8yfo**
this will look now like the following
```bash
❯ : cargo near build non-reproducible-wasm --help
Runs on current filesystem state without many restrictions
                                                                                                                                                             
Usage: cargo-near near build non-reproducible-wasm [OPTIONS]
                                                                                                                                                             
Options:
      --locked
      --no-release
      --no-abi
      --no-embed-abi
      --no-doc
      --no-wasmopt
      --out-dir <OUT_DIR>
      --manifest-path <MANIFEST_PATH>
      --features <FEATURES>
      --no-default-features
      --color <COLOR>                  [possible values: auto, always, never]
      --env <ENV>
  -h, --help                           Print help
  -V, --version                        Print version

```
```bash
❯ : cargo near build reproducible-wasm --help
Requires `docker` config added and (git)committed to Cargo.toml, runs on clean (git)working tree state
                                                                                                                                                             
Usage: cargo-near near build reproducible-wasm [OPTIONS]
                                                                                                                                                             
Options:
      --no-locked
      --out-dir <OUT_DIR>
      --manifest-path <MANIFEST_PATH>
      --color <COLOR>                  [possible values: auto, always, never]
  -h, --help                           Print help
  -V, --version                        Print version
```
---

this is pending 
- [x] deploy cmd dispatch in current pr
- [x] logic to enforce correct `container_build_command` for newer images (ones, that will be built after release of this pr)
- [x] prettier indents in interactive help msgs in current pr
- [x] separate pr onto current pr as base branch with tests fixed as per api change (#263)
- [x] testing of this pr on a few common cases (sample-crate, factory-rust, neardevhub-contract) (in comment https://github.com/near/cargo-near/pull/262#issuecomment-2529296185)
- [x] separate pr onto current pr as base branch with `.github/workflows` change /testing (#264)
- [x] separate pr onto current pr as base branch with docs/README update (#265)
- [ ] merge

### CODE REVIEW

File: cargo-near/src/commands/build/actions/non_reproducible_wasm/mod.rs
Comment (dj8yfo) [inline cargo-near/src/commands/build/actions/non_reproducible_wasm/mod.rs:202]: this would look like the following, if someone updates `image` and `image_digest`, 
but doesn't update `container_build_command`  (which stays to be `["cargo", "near", "build"]`):

image

**Comment by dj8yfo** [comment]
# 💖 __COMMENT__: prepare comment with tests' results

has been tested on following cases: 

## sample-crate

1.  https://github.com/dj8yfo/sample_no_workspace/tree/test_interactive_command_no_docker  => sample-crate-109.testnet :heavy_check_mark: 

2.  [sample-crate-110.testnet](https://testnet.nearblocks.io/address/sample-crate-110.testnet) :heavy_check_mark:   

<details>
  <summary>sample-crate-110.testnet</summary><p>
  
```json
{
  "build_info": {
    "build_command": [
      "cargo",
      "near",
      "build",
      "non-reproducible-wasm",
      "--locked"
    ],
    "build_environment": "dj8yfo/sourcescan:0.x.x-dev-pr-262@sha256:a231d4bf975d561a06dd5357f2ac03c883e8b3b510994f3b40c9b975dcdb02ce",
    "contract_path": "",
    "source_code_snapshot": "git+https://github.com/dj8yfo/sample_no_workspace?rev=0af334a86ea56d4257e6cc199231eb6a30950e69"
  },
  "link": "https://github.com/dj8yfo/sample_no_workspace/tree/0af334a86ea56d4257e6cc199231eb6a30950e69",
  "standards": [
    {
      "standard": "nep330",
      "version": "1.2.0"
    }
  ],
  "version": "0.7.7"
}
```

</p></details>


## factories-like note

For both of the following `factory-rust` and `neardevhub-contract` the product-of-factories contracts can 
be verified by SourceScan by explicitly adding [package.metadata.near.reproducible_build](https://github.com/dj8yfo/factory-rust/blob/91dde6ef874c7e0b74535c0b8cbd5d532d0caa99/product-donation/Cargo.toml#L13-L22), which is kept in sync with [build script options](https://github.com/dj8yfo/factory-rust/blob/91dde6ef874c7e0b74535c0b8cbd5d532d0caa99/factory/build.rs#L18).

(SourceScan runs `cargo near build` of version < 0.12.0 under the hood for the passed contract)

## factory-rust

api example for build scripts (no change) - [build.rs](https://github.com/dj8yfo/factory-rust/blob/91dde6ef874c7e0b74535c0b8cbd5d532d0caa99/factory/build.rs#L9-L42) 

1. ok (same bs / image) [repro-fct-80.testnet](https://testnet.nearblocks.io/address/repro-fct-80.testnet) :heavy_check_mark: , [donation-product.repro-fct-80.testnet](https://testnet.nearblocks.io/address/donation-product.repro-fct-80.testnet) :heavy_check_mark: 

<details>
  <summary>repro-fct-80.testnet</summary><p>

```json
{
  "build_info": {
    "build_command": [
      "cargo",
      "near",
      "build",
      "non-reproducible-wasm",
      "--locked",
      "--no-default-features",
      "--features",
      "near-sdk/expensive-debug"
    ],
    "build_environment": "dj8yfo/sourcescan:0.x.x-dev-pr-262-1.83@sha256:b38d13fb8c3dfe4d8555303907f8ea47be8df7bf65ee452817af5ae0c54cdc51",
    "contract_path": "factory",
    "source_code_snapshot": "git+https://github.com/dj8yfo/factory-rust?rev=91dde6ef874c7e0b74535c0b8cbd5d532d0caa99"
  },
  "link": "https://github.com/dj8yfo/factory-rust/tree/91dde6ef874c7e0b74535c0b8cbd5d532d0caa99",
  "standards": [
    {
      "standard": "nep330",
      "version": "1.2.0"
    }
  ],
  "version": "0.1.13"
}
```

</p></details>

<details>
  <summary>donation-product.repro-fct-80.testnet</summary><p>

```json
{
  "build_info": {
    "build_command": [
      "cargo",
      "near",
      "build",
      "non-reproducible-wasm",
      "--locked",
      "--no-default-features"
    ],
    "build_environment": "dj8yfo/sourcescan:0.x.x-dev-pr-262-1.83@sha256:b38d13fb8c3dfe4d8555303907f8ea47be8df7bf65ee452817af5ae0c54cdc51",
    "contract_path": "product-donation",
    "source_code_snapshot": "git+https://github.com/dj8yfo/factory-rust?rev=91dde6ef874c7e0b74535c0b8cbd5d532d0caa99"
  },
  "link": "https://github.com/dj8yfo/factory-rust/tree/91dde6ef874c7e0b74535c0b8cbd5d532d0caa99",
  "standards": [
    {
      "standard": "nep330",
      "version": "1.2.0"
    }
  ],
  "version": "0.2.9"
}
```
  
</p></details>

2. new bs / older image: [repro-fct-81.testnet](https://testnet.nearblocks.io/address/repro-fct-81.testnet) :heavy_check_mark: ,

[donation-product.repro-fct-81.testnet](https://testnet.nearblocks.io/address/donation-product.repro-fct-81.testnet) :negative_squared_cross_mark: 

fails reproducibility verification, across breaking change boundary (builder version `cargo-near-build 0.3.2` -> builder environment version `cargo-near-build 0.3.0`), `0.3.2` is current pr, to be published as `0.4.0`

<details>
  <summary>repro-fct-81.testnet</summary><p>

```json
{
  "build_info": {
    "build_command": [
      "cargo",
      "near",
      "build",
      "--no-default-features",
      "--features",
      "near-sdk/expensive-debug"
    ],
    "build_environment": "sourcescan/cargo-near:0.11.0-rust-1.82.0@sha256:6af67e796ff7182f6b184031f9e95d792f97b66693698abc51b4630acac0bf83",
    "contract_path": "factory",
    "source_code_snapshot": "git+https://github.com/dj8yfo/factory-rust?rev=9c41cf040afb0c7c315a3b5efc14f95e0532325a"
  },
  "link": "https://github.com/dj8yfo/factory-rust/tree/9c41cf040afb0c7c315a3b5efc14f95e0532325a",
  "standards": [
    {
      "standard": "nep330",
      "version": "1.2.0"
    }
  ],
  "version": "0.1.13"
}
```
</p></details>

<details>
  <summary>donation-product.repro-fct-81.testnet</summary><p>

```json
{
  "build_info": {
    "build_command": [
      "cargo",
      "near",
      "build",
      "non-reproducible-wasm",
      "--locked",
      "--no-default-features"
    ],
    "build_environment": "sourcescan/cargo-near:0.11.0-rust-1.82.0@sha256:6af67e796ff7182f6b184031f9e95d792f97b66693698abc51b4630acac0bf83",
    "contract_path": "product-donation",
    "source_code_snapshot": "git+https://github.com/dj8yfo/factory-rust?rev=9c41cf040afb0c7c315a3b5efc14f95e0532325a"
  },
  "link": "https://github.com/dj8yfo/factory-rust/tree/9c41cf040afb0c7c315a3b5efc14f95e0532325a",
  "standards": [
    {
      "standard": "nep330",
      "version": "1.2.0"
    }
  ],
  "version": "0.2.9"
}
```
  
</p></details>

<details>
  <summary>repro-fct-product-81.testnet</summary><p>

```json
{
  "build_info": {
    "build_command": [
      "cargo",
      "near",
      "build",
      "--no-default-features"
    ],
    "build_environment": "sourcescan/cargo-near:0.11.0-rust-1.82.0@sha256:6af67e796ff7182f6b184031f9e95d792f97b66693698abc51b4630acac0bf83",
    "contract_path": "product-donation",
    "source_code_snapshot": "git+https://github.com/dj8yfo/factory-rust?rev=9c41cf040afb0c7c315a3b5efc14f95e0532325a"
  },
  "link": "https://github.com/dj8yfo/factory-rust/tree/9c41cf040afb0c7c315a3b5efc14f95e0532325a",
  "standards": [
    {
      "standard": "nep330",
      "version": "1.2.0"
    }
  ],
  "version": "0.2.9"
}
```
  
</p></details>

3. old bs / newer image: [repro-fct-82.testnet](https://testnet.nearblocks.io/address/repro-fct-82.testnet) :heavy_check_mark: , 
[donation-product.repro-fct-82.testnet](https://testnet.nearblocks.io/address/donation-product.repro-fct-82.testnet) :negative_squared_cross_mark:

fails reproducibility verification, across breaking change boundary (builder version `cargo-near-build 0.3.0` -> builder environment version `cargo-near-build 0.3.2`), `0.3.2` is current pr, to be published as `0.4.0`

<details>
  <summary>repro-fct-82.testnet</summary><p>

```json
{
  "build_info": {
    "build_command": [
      "cargo",
      "near",
      "build",
      "non-reproducible-wasm",
      "--locked",
      "--no-default-features",
      "--features",
      "near-sdk/expensive-debug"
    ],
    "build_environment": "dj8yfo/sourcescan:0.x.x-dev-pr-262-1.83@sha256:b38d13fb8c3dfe4d8555303907f8ea47be8df7bf65ee452817af5ae0c54cdc51",
    "contract_path": "factory",
    "source_code_snapshot": "git+https://github.com/dj8yfo/factory-rust?rev=f9d5b4a3b3e5d17b44cbac4aa952f364fd5a6f68"
  },
  "link": "https://github.com/dj8yfo/factory-rust/tree/f9d5b4a3b3e5d17b44cbac4aa952f364fd5a6f68",
  "standards": [
    {
      "standard": "nep330",
      "version": "1.2.0"
    }
  ],
  "version": "0.1.13"
}
```
</p></details>


<details>
  <summary>donation-product.repro-fct-82.testnet</summary><p>

```json
{
  "build_info": {
    "build_command": [
      "cargo",
      "near",
      "build",
      "--no-default-features"
    ],
    "build_environment": "dj8yfo/sourcescan:0.x.x-dev-pr-262-1.83@sha256:b38d13fb8c3dfe4d8555303907f8ea47be8df7bf65ee452817af5ae0c54cdc51",
    "contract_path": "product-donation",
    "source_code_snapshot": "git+https://github.com/dj8yfo/factory-rust?rev=f9d5b4a3b3e5d17b44cbac4aa952f364fd5a6f68"
  },
  "link": "https://github.com/dj8yfo/factory-rust/tree/f9d5b4a3b3e5d17b44cbac4aa952f364fd5a6f68",
  "standards": [
    {
      "standard": "nep330",
      "version": "1.2.0"
    }
  ],
  "version": "0.2.9"
}
```
</p></details>

 
<details>
  <summary>repro-fct-product-82.testnet</summary><p>

```json
{
  "build_info": {
    "build_command": [
      "cargo",
      "near",
      "build",
      "non-reproducible-wasm",
      "--locked",
      "--no-default-features"
    ],
    "build_environment": "dj8yfo/sourcescan:0.x.x-dev-pr-262-1.83@sha256:b38d13fb8c3dfe4d8555303907f8ea47be8df7bf65ee452817af5ae0c54cdc51",
    "contract_path": "product-donation",
    "source_code_snapshot": "git+https://github.com/dj8yfo/factory-rust?rev=f9d5b4a3b3e5d17b44cbac4aa952f364fd5a6f68"
  },
  "link": "https://github.com/dj8yfo/factory-rust/tree/f9d5b4a3b3e5d17b44cbac4aa952f364fd5a6f68",
  "standards": [
    {
      "standard": "nep330",
      "version": "1.2.0"
    }
  ],
  "version": "0.2.9"
}
```
</p></details>

## neardevhub


1. no `passed_env` passed: [community.devhub-inta.testnet](https://testnet.nearblocks.io/address/community.devhub-inta.testnet) :heavy_check_mark: ,
 [uniquehandle.community.devhub-inta.testnet](https://testnet.nearblocks.io/address/uniquehandle.community.devhub-inta.testnet) :heavy_check_mark: , 
[discussions.uniquehandle.community.devhub-inta.testnet](https://testnet.nearblocks.io/address/discussions.uniquehandle.community.devhub-inta.testnet) :heavy_check_mark:

<details>
  <summary>community.devhub-inta.testnet</summary><p>

```json
{
  "build_info": {
    "build_command": [
      "cargo",
      "near",
      "build",
      "non-reproducible-wasm",
      "--locked"
    ],
    "build_environment": "dj8yfo/sourcescan:0.x.x-dev-pr-262-1.83@sha256:b38d13fb8c3dfe4d8555303907f8ea47be8df7bf65ee452817af5ae0c54cdc51",
    "contract_path": "community-factory",
    "source_code_snapshot": "git+https://github.com/dj8yfo/neardevhub-contract?rev=2716b484b35dbe1deed8d5dbfb9ac39f5bec3fed"
  },
  "link": "https://github.com/dj8yfo/neardevhub-contract/tree/2716b484b35dbe1deed8d5dbfb9ac39f5bec3fed",
  "standards": [
    {
      "standard": "nep330",
      "version": "1.2.0"
    }
  ],
  "version": "0.1.0"
}
```
</p></details>


<details>
  <summary>uniquehandle.community.devhub-inta.testnet</summary><p>

```json
{
  "build_info": {
    "build_command": [
      "cargo",
      "near",
      "build",
      "non-reproducible-wasm",
      "--locked"
    ],
    "build_environment": "dj8yfo/sourcescan:0.x.x-dev-pr-262-1.83@sha256:b38d13fb8c3dfe4d8555303907f8ea47be8df7bf65ee452817af5ae0c54cdc51",
    "contract_path": "community",
    "source_code_snapshot": "git+https://github.com/dj8yfo/neardevhub-contract?rev=2716b484b35dbe1deed8d5dbfb9ac39f5bec3fed"
  },
  "link": "https://github.com/dj8yfo/neardevhub-contract/tree/2716b484b35dbe1deed8d5dbfb9ac39f5bec3fed",
  "standards": [
    {
      "standard": "nep330",
      "version": "1.2.0"
    }
  ],
  "version": "0.1.0"
}
```
</p></details>


<details>
  <summary>discussions.uniquehandle.community.devhub-inta.testnet</summary><p>

```json
{
  "build_info": {
    "build_command": [
      "cargo",
      "near",
      "build",
      "non-reproducible-wasm",
      "--locked"
    ],
    "build_environment": "dj8yfo/sourcescan:0.x.x-dev-pr-262-1.83@sha256:b38d13fb8c3dfe4d8555303907f8ea47be8df7bf65ee452817af5ae0c54cdc51",
    "contract_path": "discussions",
    "source_code_snapshot": "git+https://github.com/dj8yfo/neardevhub-contract?rev=2716b484b35dbe1deed8d5dbfb9ac39f5bec3fed"
  },
  "link": "https://github.com/dj8yfo/neardevhub-contract/tree/2716b484b35dbe1deed8d5dbfb9ac39f5bec3fed",
  "standards": [
    {
      "standard": "nep330",
      "version": "1.2.0"
    }
  ],
  "version": "0.1.0"
}
```
</p></details>

2. 2 vars passes into `passed_env`: [community.devhub-intb.testnet](https://testnet.nearblocks.io/address/community.devhub-intb.testnet) :negative_squared_cross_mark:

**Comment by race-of-sloths** [comment]
@dj8yfo Thank you for your contribution! Your pull request is now a part of the [Race of Sloths](https://race-of-sloths.com)!
We are launching a survey "Who is faster? You vs SpaceX"

[<picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://race-of-sloths.com/images/bot-banner-dark.svg">
    <source media="(prefers-color-scheme: light)" srcset="https://race-of-sloths.com/images/bot-banner-light.svg">
    Shows inviting banner with latest news.
</picture>](https://race-of-sloths.com/latest-news)

[<picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://badge.race-of-sloths.com/dj8yfo?type=bot&pr=near/cargo-near/262&theme=dark">
    <source media="(prefers-color-scheme: light)" srcset="https://badge.race-of-sloths.com/dj8yfo?type=bot&pr=near/cargo-near/262&theme=light">
    Shows profile picture for the author of the PR
</picture>](https://race-of-sloths.com/profile/dj8yfo)


<details><summary>Current status: <i>waiting for scoring</i></summary>

We're waiting for maintainer to score this pull request with `@race-of-sloths score [0,1,2,3,5,8,13]` command. Alternatively, autoscoring [1,2] will be applied for this pull request

</details><details><summary>What is the Race of Sloths</summary>
<p>
Race of Sloths is a friendly competition where you can participate in challenges and compete with other open-source contributors within your normal workflow

For contributors:

- Tag @race-of-sloths inside your pull requests
- Wait for the maintainer to review and score your pull request
- Check out your position in the [Leaderboard](https://race-of-sloths.com/leaderboard)
- Keep weekly and monthly streaks to reach higher positions
- Boast your contributions with a dynamic picture of your [Profile](https://race-of-sloths.com/profile/dj8yfo)

For maintainers:
- Score pull requests that participate in the Race of Sloths and receive a reward
- Engage contributors with fair scoring and fast responses so they keep their streaks
- Promote the Race to the point where the Race starts promoting you
- Grow the community of your contributors

Feel free to check [our website](https://race-of-sloths.com) for additional details!
</p>
</details> 

<details>
<summary>Bot commands</summary>

- For contributors
  - **Include a PR:** `@race-of-sloths include` to enter the Race with your PR
- For maintainers:
  - **Invite contributor** `@race-of-sloths invite` to invite the contributor to participate in a race or include it, if it's already a runner.
  - **Assign points:** `@race-of-sloths score [1/2/3/5/8/13]` to award points based on your assessment.
  - **Reject this PR:** `@race-of-sloths exclude` to send this PR back to the drawing board.
  - **Exclude repo:** `@race-of-sloths pause` to stop bot activity in this repo until `@race-of-sloths unpause` command is called

</details>

File: cargo-near/src/commands/abi/mod.rs
Comment (frol) [inline cargo-near/src/commands/abi/mod.rs:7]: "implicit" looks unnecessary and confusing

```suggestion
    /// enable `--locked` flag for all `cargo` commands, disabled by default
```

File: cargo-near/src/commands/build/actions/non_reproducible_wasm/mod.rs
Comment (frol) [inline cargo-near/src/commands/build/actions/non_reproducible_wasm/mod.rs:7]:
```suggestion
    /// enable `--locked` flag for all `cargo` commands, disabled by default
```

File: cargo-near/src/commands/build/mod.rs
Comment (frol) [inline cargo-near/src/commands/build/mod.rs:40]: I think we need to be more direct for the users to be able to make the right decision (for dev vs for prod):

```suggestion
        #[strum_discriminants(strum(
            message = "non-reproducible-wasm  - Fast and simple (recommended for use during local development)"
        ))]
        /// Fast and simple (recommended for use during local development)
        NonReproducibleWasm(self::non_reproducible_wasm::BuildOpts),
        #[strum_discriminants(strum(
            message = "reproducible-wasm      - Requires [reproducible_build] section in Cargo.toml, and all changes committed and pushed to git (recommended for the production release)"
        ))]
        /// Requires [reproducible_build] section in Cargo.toml, and all changes committed and pushed to git (recommended for the production release)
```

File: cargo-near/src/commands/deploy/mod.rs
Comment (frol) [inline cargo-near/src/commands/deploy/mod.rs:20]: Apply the same reasoning as for the build commands ^

File: README.md
Comment (frol) [inline README.md:117]: Let's explicitly recommend non-reproducible-wasm for local development and reproducible-wasm for production releases

**Review by frol — CHANGES_REQUESTED** [review]

**Review by frol — APPROVED** [review]
Amazing!

@race-of-sloths score 13

**Pull request merged**
